File: js/src/ort/decoder.ts
import { COCO_CLASSES } from "../core/classes";
import { decodeDetect } from "../core/decodeRaw";
import { filterE2E } from "../core/filterE2E";
import { letterboxUnmap } from "../core/letterbox";
import { normalizeOutput } from "../core/normalize";
import type { ArrayDetections, Detection } from "../core/types";
import { canvasToImageData, imageBitmapToImageData, letterboxImageData } from "./preprocess";

export interface MinimalSession {
  inputNames: readonly string[];
  outputNames: readonly string[];
  outputMetadata?: Record<string, { dims?: readonly number[] }>;
  run(
    feeds: Record<string, { data: Float32Array; dims: readonly number[]; type: "float32" }>,
  ): Promise<Record<string, { data: Float32Array; dims: readonly number[]; type: string }>>;
}

export interface PredictOptions {
  conf?: number;
  classes?: number[];
  format?: "dict" | "arrays";
}

export type PredictInput = HTMLCanvasElement | ImageBitmap | ImageData;

export class Decoder {
  readonly backend: string;
  readonly isE2E: boolean;
  private readonly _session: MinimalSession;
  private readonly _inputName: string;
  private readonly _outputName: string;

  constructor(session: MinimalSession) {
    this._session = session;
    if (session.outputNames.length !== 1) {
      throw new Error(
        `expected single output tensor; got ${session.outputNames.length}. Is this a YOLO26 detect ONNX?`,
      );
    }
    this._inputName = session.inputNames[0] as string;
    this._outputName = session.outputNames[0] as string;
    const dims = session.outputMetadata?.[this._outputName]?.dims ?? [];
    this.isE2E = dims.length > 0 && dims[dims.length - 1] === 6;
    this.backend = "unknown";
  }

  async predict(
    input: PredictInput,
    opts: PredictOptions = {},
  ): Promise<Detection[] | ArrayDetections> {
    const { conf = 0.25, classes, format = "dict" } = opts;
    let imageData: ImageData;
    if (typeof ImageBitmap !== "undefined" && input instanceof ImageBitmap) {
      imageData = await imageBitmapToImageData(input);
    } else if (typeof HTMLCanvasElement !== "undefined" && input instanceof HTMLCanvasElement) {
      imageData = canvasToImageData(input);
    } else {
      imageData = input as ImageData;
    }

    const { tensor, meta } = letterboxImageData(imageData, 640);
    const result = await this._session.run({
      [this._inputName]: { data: tensor, dims: [1, 3, 640, 640], type: "float32" },
    });
    const raw = result[this._outputName];
    if (!raw) throw new Error(`session output ${this._outputName} missing`);
    const out = normalizeOutput(raw.data);

    let arr: ArrayDetections;
    if (this.isE2E) {
      arr = filterE2E(out, raw.dims, { conf, classes, format: "arrays" }) as ArrayDetections;
    } else {
      arr = decodeDetect(out, raw.dims, { conf, format: "arrays" }) as ArrayDetections;
    }

    if (arr.boxes.length) {
      const unmapped = letterboxUnmap(arr.boxes, meta.origSize, meta.lbSize, meta.scale, meta.pad);
      arr = { boxes: unmapped, scores: arr.scores, classes: arr.classes };
    }

    if (format === "arrays") return arr;
    const dets: Detection[] = [];
    for (let i = 0; i < arr.classes.length; i++) {
      const c = arr.classes[i] as number;
      dets.push({
        box: [
          arr.boxes[i * 4] as number,
          arr.boxes[i * 4 + 1] as number,
          arr.boxes[i * 4 + 2] as number,
          arr.boxes[i * 4 + 3] as number,
        ],
        score: arr.scores[i] as number,
        class: c,
        label: COCO_CLASSES[c] as string,
      });
    }
    return dets;
  }
}

export function fromOrt(session: MinimalSession): Decoder {
  return new Decoder(session);
}

File: python/src/yolo26_kit/ort/decoder.py
"""Auto-routing wrapper around onnxruntime.InferenceSession."""
from __future__ import annotations

from collections.abc import Iterable
from pathlib import Path
from typing import Any, Literal

from ..core.decode_raw import decode_detect
from ..core.filter_e2e import filter_e2e
from ..core.letterbox import letterbox_unmap
from ..core.normalize import normalize_output
from ..core.types import COCO_CLASSES
from .image_io import to_hwc_uint8
from .preprocess import letterbox_forward


def from_ort(session_or_path: Any, *, providers: list[str] | None = None) -> Decoder:
    if isinstance(session_or_path, (str, Path)):
        try:
            import onnxruntime as ort  # type: ignore[import-not-found]
        except ImportError as e:
            raise ImportError("install yolo26-kit[ort] to use Decoder") from e
        session = ort.InferenceSession(
            str(session_or_path),
            providers=providers or ort.get_available_providers(),
        )
    else:
        session = session_or_path
    return Decoder(session)


class Decoder:
    def __init__(self, session: Any) -> None:
        self._session = session
        outs = session.get_outputs()
        if len(outs) != 1:
            raise ValueError(
                f"expected single output tensor; got {len(outs)}. "
                "Is this a YOLO26 detect ONNX?",
            )
        self._output_name = outs[0].name
        self._input_name = session.get_inputs()[0].name
        shape = tuple(outs[0].shape)
        self.is_e2e: bool = bool(shape and shape[-1] == 6)
        provs = session.get_providers() if hasattr(session, "get_providers") else []
        self.backend: str = provs[0] if provs else "unknown"

    def predict(
        self,
        image: Any,
        conf: float = 0.25,
        classes: Iterable[int] | None = None,
        format: Literal["dict", "arrays"] = "dict",
    ) -> Any:
        hwc = to_hwc_uint8(image)
        nchw, meta = letterbox_forward(hwc, target=640)
        out = self._session.run([self._output_name], {self._input_name: nchw})[0]
        out = normalize_output(out)

        if self.is_e2e:
            dets = filter_e2e(out, conf=conf, classes=classes, format="arrays")
        else:
            dets = decode_detect(out, conf=conf, classes=classes, format="arrays")

        assert isinstance(dets, dict)
        boxes = dets["boxes"]
        if boxes.size:
            boxes = letterbox_unmap(
                boxes,
                orig_size=meta["orig_size"],
                lb_size=meta["lb_size"],
                scale=meta["scale"],
                pad=meta["pad"],
            )
        if format == "arrays":
            return {"boxes": boxes, "scores": dets["scores"], "classes": dets["classes"]}

        return [
            {
                "box": [float(b[0]), float(b[1]), float(b[2]), float(b[3])],
                "score": float(s),
                "class": int(c),
                "label": COCO_CLASSES[int(c)],
            }
            for b, s, c in zip(boxes, dets["scores"], dets["classes"], strict=True)
        ]
